### PR TITLE
Functions for simplicial complexes from polymake

### PR DESCRIPTION
**Pull request by antonydellavecchia**
Adds connected_sum and deletion

### CODE REVIEW

File: src/Combinatorics/SimplicialComplexes.jl [hunk 1]
```diff
@@ -570,3 +570,56 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+julia> is_manifold(surface_genus_2)
+true
+```
+"""
+function connected_sum(K1::SimplicialComplex, K2::SimplicialComplex, f1::Int=0, f2::Int=0)
+  return SimplicialComplex(Polymake.topaz.connected_sum(pm_object(K1), pm_object(K2), f1, f2))
+end
+
+@doc raw"""
+     deletion(K::SimplicialComplex, face::Set{Int})
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
    deletion(K::SimplicialComplex, face::Union{<:AbstractSet{Int},<:AbstractVector{Int}})
```
Does it need to be a `Set` ?

File: src/Combinatorics/SimplicialComplexes.jl [hunk 2]
```diff
@@ -570,3 +570,56 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+
+Remove the given face and all the faces containing it. 
+
+# Examples
+```jldoctest
+julia> K = simpicial_complex([[1, 2, 3], [2, 3, 4]]);
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
julia> K = simplicial_complex([[1, 2, 3], [2, 3, 4]]);
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 3]
```diff
@@ -570,3 +570,56 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
   return Polymake.topaz.isomorphic(pm_object(K1), pm_object(K2))::Bool
 end
 
+###############################################################################
+### helpful functions
+###############################################################################
+
+@doc raw"""
+     connected_sum(K1::SimplicialComplex, K2::SimplicialComplex, f1::Int=0, f2::Int=0)
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
    connected_sum(K1::SimplicialComplex, K2::SimplicialComplex, f1::Int=0, f2::Int=0)
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 4]
```diff
@@ -570,3 +570,56 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+0-th facet of both. The vertices in the selected facets are identified with each
+other according to their order in the facet (that is, in icreasing index order).
+
+# Examples
+```jldoctest
+julia> K1 = torus();
+
+julia> K2 = torus();
+
+julia> surface_genus_2 = connected(K1, K2)
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
julia> K = torus();

julia> surface_genus_2 = connected(K, K)
Abstract simplicial complex of dimension 2 on 4 vertices
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 5]
```diff
@@ -570,3 +570,56 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+julia> facets(K_with_deletion)
+2-element Vector{Set{Int64}}:
+ Set([3, 1])
+ Set([4, 2, 3])
+```
+"""
+function deletion(K::SimplicialComplex, face::Set{Int})
+  pm_set = Polymake.convert_to_pm_type(Set{Int})
+  pm_face = convert(pm_set, Polymake.to_zero_based_indexing(face))
+  return SimplicialComplex(Polymake.topaz.deletion(Oscar.pm_object(K), pm_face))
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
function deletion(K::SimplicialComplex, face::Union{<:AbstractSet{Int},<:AbstractVector{Int}})
  zero_based = Polymake.to_zero_based_indexing(face)
  return SimplicialComplex(Polymake.topaz.deletion(pm_object(K), zero_based))
```
I think allowing a vector is also fine and makes it easier to use, polymake seems to be fine with it as well. The following do work fine, so there is no need for an explicit conversion to a polymake object:
```
Polymake.topaz.deletion(Oscar.pm_object(K), Polymake.to_zero_based_indexing([1,2]))
Polymake.topaz.deletion(Oscar.pm_object(K), Polymake.to_zero_based_indexing(Set([1,2])))
```

File: src/Combinatorics/SimplicialComplexes.jl (hunk 4 above)
Comment (antonydellavecchia) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
julia> surface_genus_2 = connected_sum(K1, K2)
```

Comment (antonydellavecchia) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
julia> surface_genus_2 = connected_sum(K1, K2)
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 6]
```diff
@@ -570,3 +570,54 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+Compute the connected sum of two complexes. Parameters f1 and f2 specify which facet
+ of the first and second complex correspondingly are glued together. Default is the
+0-th facet of both. The vertices in the selected facets are identified with each
+other according to their order in the facet (that is, in icreasing index order).
+
+# Examples
+```jldoctest
+julia> K = torus();
+
+julia> surface_genus_2 = connected(K, K)
```
Comment (antonydellavecchia) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
julia> surface_genus_2 = connected_sum(K, K)
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 7]
```diff
@@ -570,3 +570,54 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+ of the first and second complex correspondingly are glued together. Default is the
+0-th facet of both. The vertices in the selected facets are identified with each
+other according to their order in the facet (that is, in icreasing index order).
+
+# Examples
+```jldoctest
+julia> K = torus();
+
+julia> surface_genus_2 = connected_sum(K, K)
+Abstract simplicial complex of dimension 2 on 4 vertices
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
Abstract simplicial complex of dimension 2 on 11 vertices
```
sorry

File: src/Combinatorics/SimplicialComplexes.jl [hunk 8]
```diff
@@ -570,3 +570,54 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+### helpful functions
+###############################################################################
+
+@doc raw"""
+    connected_sum(K1::SimplicialComplex, K2::SimplicialComplex, f1::Int=0, f2::Int=0)
+
+Compute the connected sum of two complexes. Parameters f1 and f2 specify which facet
+ of the first and second complex correspondingly are glued together. Default is the
+0-th facet of both. The vertices in the selected facets are identified with each
+other according to their order in the facet (that is, in icreasing index order).
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
other according to their order in the facet (that is, in increasing index order).
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 9]
```diff
@@ -570,3 +570,54 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
+```
+"""
+function connected_sum(K1::SimplicialComplex, K2::SimplicialComplex, f1::Int=0, f2::Int=0)
+  return SimplicialComplex(Polymake.topaz.connected_sum(pm_object(K1), pm_object(K2), f1, f2))
+end
+
+@doc raw"""
+    deletion(K::SimplicialComplex, face::Union{<:AbstractSet{Int},<:AbstractVector{Int}})
+
+Remove the given face and all the faces containing it.
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
Remove the given face and all the faces containing it from an abstract simplicial complex `K`.
```

File: src/Combinatorics/SimplicialComplexes.jl [hunk 10]
```diff
@@ -570,3 +570,54 @@ function is_isomorphic(K1::SimplicialComplex, K2::SimplicialComplex)
...
 end
 
+###############################################################################
+### helpful functions
+###############################################################################
+
+@doc raw"""
+    connected_sum(K1::SimplicialComplex, K2::SimplicialComplex, f1::Int=0, f2::Int=0)
+
+Compute the connected sum of two complexes. Parameters f1 and f2 specify which facet
```
Comment (benlorenz) [inline src/Combinatorics/SimplicialComplexes.jl]:
```suggestion
Compute the connected sum of two abstract simplicial complexes. Parameters `f1` and `f2` specify which facet
```

**Comment by thofma** [comment]
Maybe `delete` instead of `deletion`? I have a feeling we lean more towards verbs, e.g. `push!`, `delete!` etc. Just a random comment. Feel free to ignore it.

**Comment by antonydellavecchia** [comment]
> Maybe `delete` instead of `deletion`? I have a feeling we lean more towards verbs, e.g. `push!`, `delete!` etc. Just a random comment. Feel free to ignore it.

So I am just copying the name from polymake, but I think the distinction is important here. We not only delete the face but the faces containing it. We also use the same wording for matroids.

**Review by benlorenz — APPROVED** [review]

**Pull request merged**